File: clients/flutter/lib/src/transport/rest.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../error.dart';
import '../operations.dart';
import '../response.dart';
import 'transport.dart';

/// The default transport: a multipart `POST` to `{baseURL}/api/v1/transform`, with the
/// operations serialized into the query string and the image sent as the `image` field.
class RestTransport implements Transport {
  /// Creates a REST transport.
  const RestTransport({this.path = '/api/v1/transform', this.field = 'image'});

  /// The transform endpoint path.
  final String path;

  /// The multipart field name carrying the image.
  final String field;

  @override
  Future<TransformResponse> transform(TransformRequest request) async {
    final resolved = await request.input.resolve(request.httpClient);

    final query = Operations.toQuery(request.operations);
    final endpoint = '${request.baseUrl}$path${query.isEmpty ? '' : '?$query'}';

    final multipart = http.MultipartRequest('POST', Uri.parse(endpoint))
      ..headers.addAll(request.headers)
      ..files.add(http.MultipartFile.fromBytes(
        field,
        resolved.bytes,
        filename: resolved.filename,
      ));

    final http.Response response;
    try {
      final streamed =
          await request.httpClient.send(multipart).timeout(request.timeout);
      response = await http.Response.fromStream(streamed);
    } catch (error) {
      throw SharptownError('HTTP request failed: $error');
    }

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw _restError(response);
    }

    return TransformResponse(
      status: response.statusCode,
      headers: response.headers,
      bytes: response.bodyBytes,
    );
  }

  static SharptownError _restError(http.Response response) {
    var message = 'Sharptown request failed with status ${response.statusCode}';
    Object? body;
    try {
      final parsed = jsonDecode(response.body);
      body = parsed;
      if (parsed is Map && parsed['error'] is String) {
        message = parsed['error'] as String;
      }
    } catch (_) {
      body = null;
    }
    return SharptownError(message, status: response.statusCode, body: body);
  }
}

File: clients/elixir/lib/sharptown/transport/rest.ex
defmodule Sharptown.Transport.REST do
  @moduledoc """
  The REST transport — a `multipart/form-data` POST to `{base_url}/api/v1/transform` with
  operations in the query string. This is the default transport.
  """

  @behaviour Sharptown.Transport

  alias Sharptown.{Error, HTTP, JSON, Operations, Response}

  @impl true
  def transform(request, opts) do
    path = Keyword.get(opts, :path, "/api/v1/transform")
    field = Keyword.get(opts, :field, "image")

    {body, content_type} = multipart(field, request.filename, request.content_type, request.bytes)
    url = endpoint(request.base_url, path, Operations.to_query(request.operations))

    case HTTP.request(:post, url, request.headers, content_type, body, request.timeout) do
      {:ok, {status, headers, response_body}} when status in 200..299 ->
        {:ok, %Response{status: status, headers: headers, body: response_body}}

      {:ok, {status, _headers, response_body}} ->
        {:error, error_for(status, response_body)}

      {:error, reason} ->
        {:error, %Error{message: "HTTP request failed: #{inspect(reason)}"}}
    end
  end

  defp error_for(status, body) do
    message =
      case JSON.decode(body) do
        {:ok, %{"error" => error}} when is_binary(error) -> error
        _ -> "Sharptown request failed with status #{status}"
      end

    parsed = with {:ok, decoded} <- JSON.decode(body), do: decoded
    %Error{message: message, status: status, body: parsed}
  end

  defp multipart(field, filename, content_type, bytes) do
    boundary =
      "----SharptownBoundary" <> Base.encode16(:crypto.strong_rand_bytes(16), case: :lower)

    name = String.replace(filename, ["\"", "\r", "\n"], "")

    body =
      "--#{boundary}\r\n" <>
        ~s(Content-Disposition: form-data; name="#{field}"; filename="#{name}"\r\n) <>
        "Content-Type: #{content_type}\r\n\r\n" <>
        bytes <>
        "\r\n--#{boundary}--\r\n"

    {body, "multipart/form-data; boundary=#{boundary}"}
  end

  defp endpoint(base_url, path, ""), do: Sharptown.URL.http_base(base_url) <> path

  defp endpoint(base_url, path, query),
    do: Sharptown.URL.http_base(base_url) <> path <> "?" <> query
end
